File: Python/quick_sort.py
import random

def quickSort(arr, start, end):
    if start == end:
        return

    i = start
    j = end
    pivot_idx = random.randint(i, j)
    pivot = arr[pivot_idx]

    while i < j:
        while arr[i] < pivot:
            i += 1
        while arr[j] > pivot:
            j -= 1

        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

    if start < j:
        quickSort(arr, start, j)

    if end > i:
        quickSort(arr, i, end)
    
    return arr

A=[5, -1, 3, 2, -4, 2, 8, 1, 0, -7, 9, 6, 1, 4]

print(A)
A = quickSort(A, 0, len(A)-1)
print(A)

File: C++/quick_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

template <typename T>
vector<T> quickSort(const vector<T>& arr) {
    if(arr.size() < 2)
        return arr;
    
    const T* pivot = &arr.front() + arr.size() / 2 - 1;
    vector<T> lessTP;
    vector<T> greaterTP;

    for(const T* item = &arr.front(); item <= &arr.back(); item++){
        if(item == pivot) continue;
        if(*item <= *pivot) lessTP.push_back(*item);
        else greaterTP.push_back(*item);
    }

    vector<T> sorted_lessTP = quickSort(lessTP);
    vector<T> sorted_greaterTP = quickSort(greaterTP);

    sorted_lessTP.push_back(*pivot);
    sorted_lessTP.insert(sorted_lessTP.end(), sorted_greaterTP.begin(), sorted_greaterTP.end());

    return sorted_lessTP;
}

int main() {
    vector<int> arr = {5, 12, 3, 4, 7, 1, 0, 6, 19, 8, 13, 4, 2, 10, 16};
   
    cout << "input: ";
    for(int num: arr){
        cout << num << " ";
    }    
    cout << endl;

    vector<int> sorted = quickSort(arr);

    cout << "Output: ";
    for(int num: sorted){
        cout << num << " ";
    }    
    cout << endl;

    return 0;
}
/*
**output**
input: 5 12 3 4 7 1 0 6 19 8 13 4 2 10 16
Output: 0 1 2 3 4 4 5 6 7 8 10 12 13 16 19
*/